### PR TITLE
[5161] How to upgrade from ICE to XB

### PR DESCRIPTION
**Pull request by rart**
craftercms/craftercms#5161

### CODE REVIEW

File: source/developers/experience-builder.rst [hunk 1]
```diff
@@ -354,7 +355,7 @@ The complete FreeMarker template for the carousel component becomes:
 
 .. TODO Speak about the ice support classes, event capture overlay and special treatment for empty zones
 
-.. _macros:
+.. _xbMacros:
```
Comment (rart) [inline source/developers/experience-builder.rst:358]: @alhambrav I've renamed this ref (macros was way too vague). I did try to find references to it and found none but I don't know rst that well to know if there are syntaxes I might have not checked. Please help me make sure all references are updated appropriately.

File: source/system-administrators/index.rst [hunk 2]
```diff
@@ -180,6 +180,7 @@ Crafter Studio
    studio/create-site-with-link-to-remote-repo.rst
    studio/session-timeout-settings.rst
    studio/publishing-blacklist.rst
+   upgrade/upgrading-in-context-editing.rst
```
Comment (rart) [inline source/system-administrators/index.rst:183]: @alhambrav @sumerjabri please include in the right index. I used vita's suggestion of adding under sys-admin/upgrade but maybe it fits better elsewhere — particularly in/with the new IA.

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+.. Intro
+
+CrafterCMS 3.x and below had mechanisms for "in-context editing" (ICE) which are now deprecated when
+upgrading to CrafterCMS 4.x. This article contains information on how to move from the 3.x ICE system
+to the 4.x Experience Builder system. If you're starting your site from scratch, please refer to the
+:ref:`Experience builder (XB) docs <experience-builder>`.
+
+.. Motivation
+
+CrafterCMS 4 has new mechanism that enable authors to edit the content directly in its preview, short-cut
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:18]: CrafterCMS 4 has `a` new mechanism that `enables`

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+.. code-block:: html
+
+    <!-- Declare an element which should show a pencil that opens
+    <section id="banner" <@studio.iceAttr iceGroup="hero"/>>...</section>
+
+    <!-- Declares a component drop target called `features_o` -->
+    <div class="features" <@studio.componentContainerAttr target="features_o" component=contentModel/>>...</div>
+
+On the new world, ICE groups are now deprecated as authors now have more granular access to fields,
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:47]: `In` the new world?

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+^^^^^^^^^^^^^
+Simple Values
+^^^^^^^^^^^^^
+
+Starting with simple values (e.g. strings printed on tags), locate each of the freemarker interpolation expressions
+(i.e. ``${contentModel.someFieldId}``). Make sure that the interpolation expression is by itself on a html tag.
+
+For example, ``By <span>${contentModel.authorName_s}</span>`` is correct since the interpolated value
+is alone in its container tag, as opposed to ``<span>By ${contentModel.authorName_s}</span>``, where
+there's a piece of text is inside the same element as the interpolation.
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:105]: there's a piece of text ~is~ inside the same element as the interpolation

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+Once you've located the interpolations and made sure each expression is alone in its element, replace
+the plain-html tag with a CrafterCMS macro.
+
+For example ``<span>${contentModel.authorName_s}</span>`` would become ``<@crafter.span>${contentModel.authorName_s}</@crafter.span>``.
+In addition to converting the plain tag to a macro, you most specify some additional metadata depending
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:111]: typo?
   you `must` specify some additional metadata

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+on the portion of the model that you're working with. Typically, interpolations refer to a field of
+the model, which is the missing piece of metadata on the latter example. Add the ``$field`` attribute
+to your model with the value being the field id of what you're printing.
+
+.. code-block:: html
+
+    <@crafter.span $field="authorName_s">${contentModel.authorName_s}</@crafter.span>
+
+Most HTML tags have an :ref:`equivalent macro <xbMacros>`, but if you happen to be using a *rare*
+or custom tag that doesn't have a macro, use you can use the ``@crafter.tag`` macro, which will print
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:121]: Typo?
   or custom tag that doesn't have a macro, ~use~ you can use the ``@crafter.tag`` macro,

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+Collections (components and repeat groups)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Moving on to collections, there are new powerful macros that cover the most typical cases and usages
+of component collections or repeat groups.
+
+Previous macros you're using to render a collection (e.g. ``componentContainerAttr``, ``componentContainerAttr``,
+``draggableComponent``) should be replaced with either :ref:`renderComponentCollection` or
+:ref:`renderRepeatGroup`. In case the new macros don't quite fit your use case, first,
+consider the reason why they don't and preferably update your app to work with these macros do. In
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:138]: preferably update your app to work with these `new` macros ~do~.

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+inside the item element representing the component.
+
+Like with other fields, there shouldn't be any intermediate markup in between the field elements and
+their item elements.
+
+----------
+JavaScript
+----------
+
+The approach to upgrade JavaScript application will depend on how you first integrated your app with ICE.
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:151]: The approach to upgrade `a` JavaScript application

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+- ``data-studio-ice-label``
+- ``data-studio-component``
+- ``data-studio-component-path``
+- ``data-studio-embedded-item-id``
+
+The new attributes you would be setting are
+
+- ``data-craftercms-model-id``: the UUID of the model you're rendering (i.e. the value of the ``objectId`` tag on the xml)
+- ``data-craftercms-model-path``: path to the model you're rendering
+- ``data-craftercms-field-id``: the id of the field withing the content model (only applicable for fields, not for models)
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:177]: Typo?
   the id of the field `within`~g~ the content model

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+- The ``reportNavigation`` method is `replaced` by :ref:`initExperienceBuilder <js-app-initExperienceBuilder>`.
+  If you are using report navigation, there's a good chance you should initialize XB manually; for that purpose,
+  you should add ``initializeInContextEditing=false`` to your ``crafter.body_bottom`` invocation (i.e.
+  ``<@crafter.body_bottom initializeInContextEditing=false />``). Once you've done that, you're responsible
+  for initializing XB programmatically on your own using ``initExperienceBuilder``. You should invoke,
+  ``initExperienceBuilder`` after each new page is rendered and you want to tell Studio it is now viewing
+  a different page. Before initializing a new view, once the navigation has occurred you should invoke
+  the unmount function that's returned by ``initExperienceBuilder`` (e.g. ``initExperienceBuilder({ ... }).unmount()``).
+
+- The new package too exports ``fetchIsAuthoring`` and ``addAuthoringSupport`` and the methods on both packages are equivalent.
```
Comment (alhambrav) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:212]: Typo?
   The new package `tool` exports

**Review by alhambrav — CHANGES_REQUESTED** [review]

**Review by alhambrav — APPROVED** [review]

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+----------
+JavaScript
+----------
+
+The approach to upgrade JavaScript applications will depend on how you first integrated your app with ICE.
+If you didn't have in-context editing on your app, head to :ref:`XB docs <xbJsApps>` to learn how to integrate.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Migrating from manual attributes
```
Comment (sumerjabri) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:155]: Fix the case to match other similar sections. Like this: `Migrating From Manual Attributes`

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+As mentioned before, the recommended approach is to use the JavaScript SDK, specifically the
+`@craftercms/experience-builder <https://www.npmjs.com/package/@craftercms/experience-builder>`_ package.
+Once you install or download the package, from its `index`, you should use :ref:`getICEAttributes` which will
+create and return an object with all the necessary attributes and values for you to simply apply the
+attributes to your elements in whatever way suits your app. If you're using React, you should use the
+:ref:`React-specific bindings <xb-react-bindings>` that the package offers.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Migrating from ``@craftercms/ice``
```
Comment (sumerjabri) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:188]: From

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  you should add ``initializeInContextEditing=false`` to your ``crafter.body_bottom`` invocation (i.e.
+  ``<@crafter.body_bottom initializeInContextEditing=false />``). Once you've done that, you're responsible
+  for initializing XB programmatically on your own using ``initExperienceBuilder``. You should invoke,
+  ``initExperienceBuilder`` after each new page is rendered and you want to tell Studio it is now viewing
+  a different page. Before initializing a new view, once the navigation has occurred you should invoke
+  the unmount function that's returned by ``initExperienceBuilder`` (e.g. ``initExperienceBuilder({ ... }).unmount()``).
+
+- The new package also exports ``fetchIsAuthoring`` and ``addAuthoringSupport`` and the methods on both packages are equivalent.
+
+----
```
Comment (sumerjabri) [inline source/system-administrators/upgrade/upgrading-in-context-editing.rst:214]: Remove this section

File: source/developers/experience-builder.rst [hunk 15]
```diff
@@ -1248,6 +1253,8 @@ Usage
 XB JS libraries can be used either via npm by importing ``@craftercms/experience-builder`` or using the
 JS umd bundle and adding it into your app's runtime.
```
Comment (sumerjabri) [inline source/developers/experience-builder.rst:1254]: What's JS umd?

File: source/site-administrators/studio/user-interface-configuration.rst [hunk 16]
```diff
@@ -276,9 +276,12 @@ Here's the Sidebar with the folder ``2021/3`` hidden:
...
 
-*****************************
-Sidebar Widget Icon/Container
-*****************************
+*********************************
+Navigator Widgets Styling Options
+*********************************
+
+Several styling options are available for the navigator widgets. Both the widget's container element
+and the icon element can receive css classes and targeted base styles, collapsed styles and expanded styles.
```
Comment (sumerjabri) [inline source/site-administrators/studio/user-interface-configuration.rst:284]: CSS

File: source/site-administrators/studio/user-interface-configuration.rst [hunk 17]
```diff
@@ -351,6 +354,26 @@ Here's the Sidebar with the colors and border added:
...
 |
 
+
+Container Options
+=================
+
+- ``baseClass``: a class name (string) applied to the container regardless of state
+- ``expandedClass``: a class name (string) applied to the container when expanded
+- ``collapsedClass``: a class name (string) applied to the container when collapsed
+- ``baseStyle``: a set of css rules applied to the container regardless of state (e.g. ``<baseStyles><backgroundColor>red</backgroundColor></baseStyles>``)
```
Comment (sumerjabri) [inline source/site-administrators/studio/user-interface-configuration.rst:364]: CSS

File: source/site-administrators/studio/user-interface-configuration.rst [hunk 18]
```diff
@@ -351,6 +354,26 @@ Here's the Sidebar with the colors and border added:
...
+- ``expandedStyle``: a set of css rules applied to the container when expanded
+- ``collapsedStyle``: a set of css rules applied to the container when collapsed
+
+Icon Options
+============
+
+- ``class``: a class name (string) applied to the icon regardless of state
+- ``style``: a set of css rules applied to the icon regardless of state  (e.g. ``<baseStyles><backgroundColor>red</backgroundColor></baseStyles>``)
```
Comment (sumerjabri) [inline source/site-administrators/studio/user-interface-configuration.rst:372]: CSS

**Review by sumerjabri — CHANGES_REQUESTED** [review]

File: source/system-administrators/index.rst (hunk 2 above)
Comment (rart) [reply source/system-administrators/index.rst:183]: Who's this action item for? If for me, I need more details to be able to execute.

File: source/system-administrators/upgrade/upgrading-in-context-editing.rst (hunk 14 above)
Comment (rart) [reply source/system-administrators/upgrade/upgrading-in-context-editing.rst:214]: I don't think this is necessary. As part of the upgrade activities devs will be removing one package and adding a new one. These methods coming from the old package may create uncertainty as to what to do with these. Don't see any harm in leaving it and it's reassurance on what the upgrade process looks like.

File: source/developers/experience-builder.rst (hunk 15 above)
Comment (rart) [reply source/developers/experience-builder.rst:1254]: A term from the JS community/ecosystem: [Universal Module Definition](https://riptutorial.com/javascript/example/16339/universal-module-definition--umd-)

File: source/site-administrators/studio/user-interface-configuration.rst (hunk 16 above)
Comment (rart) [reply source/site-administrators/studio/user-interface-configuration.rst:284]: @alhambrav @sumerjabri FYI:
- There's a few other instances around the docs where `css` is not in uppercase.
- I've just noticed that documents are duplicated under `/new-ia`; not sure what the or update/transition protocol is. Right now these two are out of sync.

File: source/system-administrators/index.rst (hunk 2 above)
Comment (sumerjabri) [reply source/system-administrators/index.rst:183]: I put an @alhambrav in the comment

**Review by sumerjabri — APPROVED** [review]
@alhambrav will move the articles to the developer's section and cite that there and in the release notes.

**Pull request merged**